flybywiresim/aircraft
# feat(CDU): return to previous FPlan position

Fixes #7422

## Summary of Changes


The primary aim of the PR is to F-PLAN offset stateful to be able to return back to same scroll.
Offset changes on these situations:
* New Departure insertion -- resets to top
* New Arrival -- resets to bottom
* Airway insertion -- moves by number of inserted waypoints
* New INIT (incl. AOC) -- resests to top
* Second press of FPLAN button -- resets to top

The PR also refactors FlightPlanManager to await flight plan version updates. This is to ensure that a new offset set on waypoint insertions is correct, as it is computed and limited to actual waypoint numbers.

## Screenshots (if necessary)



## References

https://youtu.be/syr4G3pokgU?t=191
https://youtu.be/_skUOMfv-AQ
https://youtu.be/E8IjkgQjw2M
https://youtu.be/D7HjVlvKd8o
https://youtu.be/dDny7J9Bc1Y
https://youtu.be/tXnbLtRpcqM





## Additional context



Discord username (if different from GitHub): coolll.sk#0022

## Testing instructions


For FPLAN:
* validate returns from sub pages returning to FPLAN -- scroll should be the same
  * Departure, Arrival ... (with no changes)
  * Vertical and Lateral revision pages
  * Holds
* validate return after insertions from Departure, Arrival, Airways
  * New Departure insertion -- resets to top
  * New Arrival -- resets to bottom
  * Airway insertion -- moves by number of inserted waypoints
* validate second press of FPLAN button -- resets to top
* regression test AIRPORT button 

For the FPM refactoring:
* regression testing with a full flight
* observe INIT from AOC
* observe DIR to behavior
* try waypoint changes during curise


## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page
